Expose form requestURL in _WKInputDelegate willSubmitFormValues callback

arenevier opened the pull request:
#### ce773e32edaffb27f60e22fc2588041cc5884778
<pre>
Expose form requestURL in _WKInputDelegate willSubmitFormValues callback
<a href="https://bugs.webkit.org/show_bug.cgi?id=302341">https://bugs.webkit.org/show_bug.cgi?id=302341</a>
<a href="https://rdar.apple.com/164485992">rdar://164485992</a>

Reviewed by Alex Christensen and BJ Burg.

Send form requestURL and method to UIProcess. Add a new method to
_WKInputDelegate that exposes those parameters.

Instead of passing around the FormState, we now pass the
FormSubmission, and it will be used in get the requestURL and method in
FrameLoader::continueLoadAfterNavigationPolicy

Tests: Tools/TestWebKitAPI/Tests/WebKitCocoa/DuplicateCompletionHandlerCalls.mm
       Tools/TestWebKitAPI/Tests/WebKitCocoa/UserContentController.mm
       Tools/TestWebKitAPI/Tests/WebKitCocoa/_WKInputDelegate.mm

Tests: Tools/TestWebKitAPI/Tests/WebKitCocoa/DuplicateCompletionHandlerCalls.mm
       Tools/TestWebKitAPI/Tests/WebKitCocoa/UserContentController.mm
       Tools/TestWebKitAPI/Tests/WebKitCocoa/_WKInputDelegate.mm

* Source/WebCore/html/HTMLFormElement.h:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::willSendRequest):
* Source/WebCore/loader/DocumentLoader.h:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyFrameLoaderClient::dispatchWillSubmitForm):
* Source/WebCore/loader/EmptyFrameLoaderClient.h:
* Source/WebCore/loader/FormSubmission.h:
(WebCore::FormSubmission::formState const):
(WebCore::FormSubmission::takeState): Deleted.
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::loadFrameRequest):
(WebCore::FrameLoader::loadURL):
(WebCore::FrameLoader::load):
(WebCore::FrameLoader::loadWithNavigationAction):
(WebCore::FrameLoader::loadWithDocumentLoader):
(WebCore::FrameLoader::loadPostRequest):
(WebCore::FrameLoader::continueLoadAfterNavigationPolicy):
(WebCore::FrameLoader::continueLoadAfterNewWindowPolicy):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebCore/loader/NavigationScheduler.cpp:
* Source/WebCore/loader/PolicyChecker.cpp:
(WebCore::PolicyChecker::checkNavigationPolicy):
(WebCore::PolicyChecker::checkNewWindowPolicy):
* Source/WebCore/loader/PolicyChecker.h:
* Source/WebKit/UIProcess/API/APIFormClient.h:
(API::FormClient::willSubmitForm):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setInputDelegate:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKInputDelegate.h:
* Source/WebKit/UIProcess/API/glib/WebKitFormClient.cpp:
* Source/WebKit/UIProcess/WebFormClient.cpp:
(WebKit::WebFormClient::willSubmitForm):
* Source/WebKit/UIProcess/WebFormClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::willSubmitForm):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::dispatchWillSubmitForm):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(WebFrameLoaderClient::dispatchWillSubmitForm):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/DuplicateCompletionHandlerCalls.mm:
(-[DuplicateCompletionHandlerCallsDelegate _webView:willSubmitFormValues:frameInfo:sourceFrameInfo:userObject:requestURL:method:submissionHandler:]):
(-[DuplicateCompletionHandlerCallsDelegate _webView:willSubmitFormValues:frameInfo:sourceFrameInfo:userObject:submissionHandler:]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UserContentController.mm:
(-[InputDelegateForFormSubmission _webView:willSubmitFormValues:frameInfo:sourceFrameInfo:userObject:requestURL:method:submissionHandler:]):
(-[InputDelegateForFormSubmission _webView:willSubmitFormValues:frameInfo:sourceFrameInfo:userObject:submissionHandler:]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/_WKInputDelegate.mm:
(-[InputDelegate _webView:willSubmitFormValues:frameInfo:sourceFrameInfo:userObject:requestURL:method:submissionHandler:]):
(-[InputDelegate _webView:willSubmitFormValues:frameInfo:sourceFrameInfo:userObject:submissionHandler:]): Deleted.

Canonical link: <a href="https://commits.webkit.org/303302@main">https://commits.webkit.org/303302@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ebacd97a21a563e41e4bb4c287b013a83b86ee3c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131969 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4461 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42983 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139482 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/83866 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ce330b6f-036b-4e0e-a418-ed456623f290) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133839 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4403 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4223 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100874 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/83866 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/41738d83-561d-4261-9c8a-71cf6227bd10) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134915 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118187 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81665 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d3954497-b956-43ff-984b-fd9af9a2b114) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3026 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82701 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111778 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36311 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142127 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4131 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36891 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109245 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4212 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/3606 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109415 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27711 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3125 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114467 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57352 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4184 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32864 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4016 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67631 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/4276 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4144 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

lilyspiniolas commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -4195,7 +4195,13 @@ void FrameLoader::continueLoadAfterNavigationPolicy(const ResourceRequest& reque
 
-    m_client->dispatchWillSubmitForm(*formState, WTFMove(completionHandler));
+    Ref form = formState->form();
+    auto* submitter = formState->submitter();
```
```suggestion
    RefPtr submitter = formState->submitter();
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitFormClient.cpp:
```diff
@@ -37,7 +37,7 @@ class FormClient final : public API::FormClient {
 
 private:
-    void willSubmitForm(WebPageProxy&, WebFrameProxy&, WebFrameProxy&, WebKit::FrameInfoData&&, WebKit::FrameInfoData&&, const Vector<std::pair<String, String>>& values, API::Object*, CompletionHandler<void()>&& completionHandler) override
+    void willSubmitForm(WebPageProxy&, WebFrameProxy&, WebFrameProxy&, WebKit::FrameInfoData&&, WebKit::FrameInfoData&&, const Vector<std::pair<String, String>>& values, API::Object*, const WTF::URL&& requestURL, const WTF::String&& method, CompletionHandler<void()>&& completionHandler) override
```
WebKit layer does not use -Wunused-parameter to avoid the need for using `UNUSED_PARAM()` all over the place, which is why I expect this to survive EWS. But it's still nicer to not name parameters that are not used.

burg replied:
I could go either way here, since URL and String are very generic types.

burg commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -5844,15 +5849,28 @@ void willSubmitForm(WebKit::WebPageProxy& page, WebKit::WebFrameProxy&, WebKit::
             }
 
-            auto checker = WebKit::CompletionHandlerCallChecker::create(inputDelegate.get(), @selector(_webView:willSubmitFormValues:frameInfo:sourceFrameInfo:userObject:submissionHandler:));
+            if (inputDelegateResponseToWillSubmitFormValuesWithoutRequestURL) {
```
Spelling: inputDelegate**Responds**To...

burg replied:
Since we'd expect to remove this older delegate method after moving over to the new method, I think you should add a FIXME to that effect.

If the 'legacy' delegate method is similarly about to be removed because its only client (Safari) moves to the new delegate method, please add a FIXME to that effect.

burg commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -5844,15 +5849,28 @@ void willSubmitForm(WebKit::WebPageProxy& page, WebKit::WebFrameProxy&, WebKit::
+                    checker->didCallCompletionHandler();
+                    completionHandler();
+                }).get()];
+            }
```
I believe you'll want to `return` at the end of this `if`. We only want to call at most one of the delegate methods.

burg commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -5811,18 +5811,23 @@ explicit FormClient(WKWebView *webView)
         virtual ~FormClient() { }
 
-        void willSubmitForm(WebKit::WebPageProxy& page, WebKit::WebFrameProxy&, WebKit::WebFrameProxy&, WebKit::FrameInfoData&& frameInfoData, WebKit::FrameInfoData&& sourceFrameInfoData, const Vector<std::pair<WTF::String, WTF::String>>& textFieldValues, API::Object* userData, CompletionHandler<void()>&& completionHandler) override
+        void willSubmitForm(WebKit::WebPageProxy& page, WebKit::WebFrameProxy&, WebKit::WebFrameProxy&, WebKit::FrameInfoData&& frameInfoData, WebKit::FrameInfoData&& sourceFrameInfoData, const Vector<std::pair<WTF::String, WTF::String>>& textFieldValues, API::Object* userData, const   WTF::URL&& requestURL, const WTF::String&& method, CompletionHandler<void()>&& completionHandler) override
```
Nit: double space

burg commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -5811,18 +5811,23 @@ explicit FormClient(WKWebView *webView)
+            SEL willSubmitFormValuesLegacySelector = @selector(_webView:willSubmitFormValues:userObject:submissionHandler:);
 
-            if (!inputDelegateResponseToWillSubmitFormValues && !inputDelegateResponseToWillSubmitFormValuesLegacy) {
+            bool inputDelegateResponseToWillSubmitFormValues = [inputDelegate respondsToSelector:willSubmitFormValuesSelector];
```
RespondsTo

burg commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/_WKInputDelegate.mm:
```diff
@@ -77,6 +77,8 @@ - (void)_webView:(WKWebView *)webView willSubmitFormValues:(NSDictionary *)value
     EXPECT_STREQ([[values objectForKey:@"testname1"] UTF8String], "testvalue1");
     EXPECT_STREQ([[values objectForKey:@"testname2"] UTF8String], "testvalue2");
+    EXPECT_STREQ([[requestURL absoluteString] UTF8String], "test:///formtarget");
+    EXPECT_STREQ([method UTF8String], "post");
```
It bothers me that this is `"post"` and not `"POST"`. Perhaps we need to add a call to `WebCore::normalizeHTTPMethod()` (in WebProcess?).

burg commented on Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
```diff
@@ -1134,7 +1135,7 @@ void WebLocalFrameLoaderClient::dispatchWillSubmitForm(FormState& formState, Com
     }
 
-    webPage->sendWithAsyncReply(Messages::WebPageProxy::WillSubmitForm(m_frame->info(), sourceFrame->info(), values, UserData(WebProcess::singleton().transformObjectsToHandles(userData.get()).get())), WTFMove(completionHandler));
+    webPage->sendWithAsyncReply(Messages::WebPageProxy::WillSubmitForm(m_frame->info(), sourceFrame->info(), values, UserData(WebProcess::singleton().transformObjectsToHandles(userData.get()).get()), requestURL, method), WTFMove(completionHandler));
```
Style: consider brace initializers:
`WillSubmitForm { m_frame->info(), sourceFrame->info(), values, UserData { ... } }`

burg commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -4195,7 +4195,13 @@ void FrameLoader::continueLoadAfterNavigationPolicy(const ResourceRequest& reque
+    auto requestURL = formSubmission->requestURL();
+    String method = FormSubmission::Attributes::methodString(formSubmission->method());
+
+    m_client->dispatchWillSubmitForm(*formState, WTFMove(requestURL), WTFMove(method), WTFMove(completionHandler));
```
Since this is v3 of this delegate method to add more data, I am a little curious whether the longer term solution is to expose an API equivalent of FormSubmission similar to _WKFrameInfo. this would allow to add more single fields without further complicating the port-specific code.

achristensen07 commented on Source/WebCore/html/HTMLFormElement.h:
```diff
@@ -121,6 +121,7 @@ class HTMLFormElement final : public HTMLElement {
     static HTMLFormElement* findClosestFormAncestor(const Element&);
     
     RefPtr<DOMFormData> constructEntryList(RefPtr<HTMLFormControlElement>&&, Ref<DOMFormData>&&, StringPairVector*);
+    FormSubmission::Attributes attributes() const { return m_attributes; }
```
const FormSubmission::Attributes& to prevent unnecessary copies.

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/_WKInputDelegate.h:
```diff
@@ -47,7 +47,8 @@ typedef NS_ENUM(NSInteger, _WKFocusStartsInputSessionPolicy) {
 - (void)_webView:(WKWebView *)webView didStartInputSession:(id <_WKFormInputSession>)inputSession;
 - (void)_webView:(WKWebView *)webView willSubmitFormValues:(NSDictionary *)values userObject:(NSObject <NSSecureCoding> *)userObject submissionHandler:(void (^)(void))submissionHandler;
-- (void)_webView:(WKWebView *)webView willSubmitFormValues:(NSDictionary *)values frameInfo:(WKFrameInfo *)frameInfo sourceFrameInfo:(WKFrameInfo *)sourceFrameInfo userObject:(NSObject <NSSecureCoding> *)userObject submissionHandler:(void (^)(void))submissionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
+- (void)_webView:(WKWebView *)webView willSubmitFormValues:(NSDictionary *)values frameInfo:(WKFrameInfo *)frameInfo sourceFrameInfo:(WKFrameInfo *)sourceFrameInfo userObject:(NSObject <NSSecureCoding> *)userObject submissionHandler:(void (^)(void))submissionHandler WK_API_DEPRECATED_WITH_REPLACEMENT("_webView:willSubmitFormValues:frameInfo:sourceFrameInfo:userObject:requestURL:method:submissionH ...
```
I don't see a strong reason to deprecate this but not the first one.  We may as well deprecate none of them right now.

burg commented on Source/WebCore/loader/FormSubmission.h:
```diff
@@ -88,7 +88,7 @@ class FormSubmission : public RefCountedAndCanMakeWeakPtr<FormSubmission> {
     const String& contentType() const { return m_contentType; }
     FormState& state() const { return *m_formState; }
-    Ref<FormState> takeState() { return m_formState.releaseNonNull(); }
+    RefPtr<FormState> refState() const { return m_formState; }
```
Nit: `protectedState()`?

achristensen07 replied:
I don't think refState is a good name.  formState would be a better name.

burg commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1475,6 +1475,7 @@ void FrameLoader::loadFrameRequest(FrameLoadRequest&& request, Event* event, Ref
     else
         loadType = FrameLoadType::Standard;
 
+    RefPtr formState = formSubmission ? formSubmission->refState(): nullptr;
```
Style nit: space before :

burg commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1568,7 +1569,7 @@ void FrameLoader::loadURL(FrameLoadRequest&& frameLoadRequest, const String& ref
     // Anchor target is ignored when the download attribute is set since it will download the hyperlink rather than follow it.
     auto effectiveFrameName = frameLoadRequest.downloadAttribute().isNull() ? frameLoadRequest.frameName() : nullAtom();
-    bool isFormSubmission = formState;
+    bool isFormSubmission = formSubmission;
```
Style nit: `!!formSubmission`

burg commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1892,7 +1894,7 @@ void FrameLoader::loadWithDocumentLoader(DocumentLoader* loader, FrameLoadType t
     policyChecker().setLoadType(type);
     RELEASE_ASSERT(!isBackForwardLoadType(type) || history().provisionalItem());
-    bool isFormSubmission = formState;
+    bool isFormSubmission = formSubmission;
```
Ditto

burg commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1650,6 +1651,7 @@ void FrameLoader::loadURL(FrameLoadRequest&& frameLoadRequest, const String& ref
 
     bool sameURL = shouldTreatURLAsSameAsCurrent(frameLoadRequest.protectedRequesterSecurityOrigin().ptr(), newURL);
     const String& httpMethod = request.httpMethod();
+    RefPtr formState = formSubmission ? formSubmission->refState(): nullptr;
```
What is this used for?

arenevier replied:
Withouth that patch, formState is passed as an argument, and is used in different places:  line 1660, line 1677, line 1679.

burg commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -3617,21 +3619,22 @@ void FrameLoader::loadPostRequest(FrameLoadRequest&& request, const String& refe
     }
 
     if (request.protectedRequesterSecurityOrigin()->isSameOriginDomain(frame->protectedDocument()->protectedSecurityOrigin().get())) {
+        RefPtr formState = formSubmission ? formSubmission->refState(): nullptr;
```
ditto

arenevier replied:
same. It's used on the line below.

achristensen07 commented on Source/WebCore/loader/DocumentLoader.cpp:
```diff
@@ -768,7 +768,7 @@ void DocumentLoader::willSendRequest(ResourceRequest&& newRequest, const Resourc
         return completionHandler(WTFMove(newRequest));
 
-    auto navigationPolicyCompletionHandler = [this, protectedThis = Ref { *this }, frame, completionHandler = WTFMove(completionHandler)] (ResourceRequest&& request, WeakPtr<FormState>&&, NavigationPolicyDecision navigationPolicyDecision) mutable {
+    auto navigationPolicyCompletionHandler = [this, protectedThis = Ref { *this }, frame, completionHandler = WTFMove(completionHandler)] (ResourceRequest&& request, const WeakPtr<FormSubmission>&&, NavigationPolicyDecision navigationPolicyDecision) mutable {
```
I don't think this is a good use of a const r-value.  Just WeakPtr&lt;FormSubmission&gt;&&

achristensen07 commented on Source/WebCore/loader/FrameLoader.h:
```diff
@@ -136,7 +136,7 @@ class FrameLoader final : public CanMakeWeakPtr<FrameLoader> {
 
     // FIXME: These are all functions which start loads. We have too many.
-    WEBCORE_EXPORT void loadFrameRequest(FrameLoadRequest&&, Event*, RefPtr<FormState>&&, std::optional<PrivateClickMeasurement>&& = std::nullopt); // Called by submitForm, calls loadPostRequest and loadURL.
+    WEBCORE_EXPORT void loadFrameRequest(FrameLoadRequest&&, Event*, RefPtr<FormSubmission>, std::optional<PrivateClickMeasurement>&& = std::nullopt); // Called by submitForm, calls loadPostRequest and loadURL.
```
Let's keep the RefPtr r-value.

achristensen07 commented on Source/WebCore/loader/FrameLoader.h:
```diff
@@ -402,8 +402,8 @@ class FrameLoader final : public CanMakeWeakPtr<FrameLoader> {
-    void continueLoadAfterNavigationPolicy(const ResourceRequest&, FormState*, NavigationPolicyDecision, AllowNavigationToInvalidURL);
-    void continueLoadAfterNewWindowPolicy(ResourceRequest&&, FormState*, const AtomString& frameName, const NavigationAction&, ShouldContinuePolicyCheck, AllowNavigationToInvalidURL, NewFrameOpenerPolicy);
+    void continueLoadAfterNavigationPolicy(const ResourceRequest&, FormSubmission*, NavigationPolicyDecision, AllowNavigationToInvalidURL);
+    void continueLoadAfterNewWindowPolicy(ResourceRequest&&, RefPtr<FormSubmission>, const AtomString& frameName, const NavigationAction&, ShouldContinuePolicyCheck, AllowNavigationToInvalidURL, NewFrameOpenerPolicy);
```
RefPtr r-value.

arenevier committed:
Expose form requestURL in _WKInputDelegate willSubmitFormValues callback
https://bugs.webkit.org/show_bug.cgi?id=302341
rdar://164485992

Reviewed by Alex Christensen and BJ Burg.

Send form requestURL and method to UIProcess. Add a new method to
_WKInputDelegate that exposes those parameters.

Instead of passing around the FormState, we now pass the
FormSubmission, and it will be used in get the requestURL and method in
FrameLoader::continueLoadAfterNavigationPolicy

Tests: Tools/TestWebKitAPI/Tests/WebKitCocoa/DuplicateCompletionHandlerCalls.mm
       Tools/TestWebKitAPI/Tests/WebKitCocoa/UserContentController.mm
       Tools/TestWebKitAPI/Tests/WebKitCocoa/_WKInputDelegate.mm

Tests: Tools/TestWebKitAPI/Tests/WebKitCocoa/DuplicateCompletionHandlerCalls.mm
       Tools/TestWebKitAPI/Tests/WebKitCocoa/UserContentController.mm
       Tools/TestWebKitAPI/Tests/WebKitCocoa/_WKInputDelegate.mm

* Source/WebCore/html/HTMLFormElement.h:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::willSendRequest):
* Source/WebCore/loader/DocumentLoader.h:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyFrameLoaderClient::dispatchWillSubmitForm):
* Source/WebCore/loader/EmptyFrameLoaderClient.h:
* Source/WebCore/loader/FormSubmission.h:
(WebCore::FormSubmission::formState const):
(WebCore::FormSubmission::takeState): Deleted.
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::loadFrameRequest):
(WebCore::FrameLoader::loadURL):
(WebCore::FrameLoader::load):
(WebCore::FrameLoader::loadWithNavigationAction):
(WebCore::FrameLoader::loadWithDocumentLoader):
(WebCore::FrameLoader::loadPostRequest):
(WebCore::FrameLoader::continueLoadAfterNavigationPolicy):
(WebCore::FrameLoader::continueLoadAfterNewWindowPolicy):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebCore/loader/NavigationScheduler.cpp:
* Source/WebCore/loader/PolicyChecker.cpp:
(WebCore::PolicyChecker::checkNavigationPolicy):
(WebCore::PolicyChecker::checkNewWindowPolicy):
* Source/WebCore/loader/PolicyChecker.h:
* Source/WebKit/UIProcess/API/APIFormClient.h:
(API::FormClient::willSubmitForm):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setInputDelegate:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKInputDelegate.h:
* Source/WebKit/UIProcess/API/glib/WebKitFormClient.cpp:
* Source/WebKit/UIProcess/WebFormClient.cpp:
(WebKit::WebFormClient::willSubmitForm):
* Source/WebKit/UIProcess/WebFormClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::willSubmitForm):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::dispatchWillSubmitForm):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(WebFrameLoaderClient::dispatchWillSubmitForm):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/DuplicateCompletionHandlerCalls.mm:
(-[DuplicateCompletionHandlerCallsDelegate _webView:willSubmitFormValues:frameInfo:sourceFrameInfo:userObject:requestURL:method:submissionHandler:]):
(-[DuplicateCompletionHandlerCallsDelegate _webView:willSubmitFormValues:frameInfo:sourceFrameInfo:userObject:submissionHandler:]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UserContentController.mm:
(-[InputDelegateForFormSubmission _webView:willSubmitFormValues:frameInfo:sourceFrameInfo:userObject:requestURL:method:submissionHandler:]):
(-[InputDelegateForFormSubmission _webView:willSubmitFormValues:frameInfo:sourceFrameInfo:userObject:submissionHandler:]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/_WKInputDelegate.mm:
(-[InputDelegate _webView:willSubmitFormValues:frameInfo:sourceFrameInfo:userObject:requestURL:method:submissionHandler:]):
(-[InputDelegate _webView:willSubmitFormValues:frameInfo:sourceFrameInfo:userObject:submissionHandler:]): Deleted.

Canonical link: https://commits.webkit.org/303302@main
